### ISSUE TITLE
Use an index for collections of facilities

### ISSUE THREAD

**Issue by ctford** [facilityregistry/fred-api#4]
Rather than include all the facility information in a collection, could we use a simple index of links and summaries to all the available facilities?

The index could use Atom, or a similar format.

This would have the following advantages:
- We separate out the two problems of listing and describing facilities, thus simplifying each representation.
- It would be easier to add new collection concepts, for example we might implement hierarchies by allowing collections to link to both individual facilities and collections of facilities, and we wouldn't have to overload further the facility format.
- Clients can load index information about facilities without having to load each facility.
- We might be able to cache the index and individual facilities separately.

Apologies if this has already been discussed.

**Comment by rowenaluk** [comment]
can we do this in addition to (as opposed to 'instead of') being able to query for all the facility information in a collection?

**Comment by ctford** [comment]
We could support both, but I guess the tradeoff is increased complexity.

What's the use case of getting all the info from the collection? Less HTTP requests?

**Comment by rowenaluk** [comment]
Seems like not too much effort. And yes, reduced HTTP requests, which I can see being a win when working with poor bandwidth scenarios.

**Comment by bobjolliffe** [comment]
The use case for getting the complete information in a collection would be
where one system is trying to synchronise its entire collection with
another.  As we discussed on the call this use case is likely to be quite
common.  In which case it would be painful, slow and inefficient to
download in index format (using the last_updated option) then iterate
through each one,

**Comment by ctford** [comment]
I'm not sure I agree with the point about an index being less efficient in terms of bandwidth. If the facilities are broken up into index and individual facilities, it potentially lessens bandwidth usage because you only have to request the things you're interested in. Separate requests is also potentially more robust in a poor connectivity situation, because if something fails you can retry just that bit.

It is, however, potentially slower because each new HTTP request introduces new latency. And I can see that it's more complicated to implement.

As I mentioned on the recent call, if bandwidth is a key concern I wonder if we should suggest that registry servers support gzip encoding if requested by a consumer? That would result in roughly 2/3 less data that needs to come across the wire. of course, there's nothing to prevent them from supporting it even if we don't mention it.

**Comment by bobjolliffe** [comment]
Yes there is a lot of "if this then that ...".  But ...
1. I agree that an index can be a very useful thing.  Amongst other
   reasons, it provides the very desirable feature of navigability.
2. In at least one known use case - DHIS2 synching its orgunits from a
   nominated facility registry (RM) in Rwanda - it wouldn't actually be much
   use.  The only requests DHIS2 would make would be a periodic (polled)
   requests for the full collection of facilities which have changed since the
   previous poll.

Caveat to 2 ... after an initial settling period it is expected that these
updates would be few and infrequent.  So the chattiness would become less
of an issue.  Mind you we are only talking of a few hundred facilities.
 Countries with 10s of 1000s of facilities would obviously see more updates.

We have sufficient functionality defined currently to meet our known use
cases so I suggest we bump the issue of index view to v1.1 and proceed more
rapidly with what we have.

**Comment by ctford** [comment]
Yes, I agree we shouldn't be expanding the scope of v1.0. We shouldn't require both the full-representation feed and an index as they really are alternatives for most use cases I can think of.

If we don't think that an index is better than a full feed, then we should park it.

I do happen to think it is better, personally, because it's cleaner to separate out enumerating and describing facilities, which I think will become even apparent if we start to have hierarchies.

**Comment by rowenaluk** [comment]
gzip sounds useful. i'd be in favor of adding that as an optional but recommended format to support, in addition to exposing the unzipped data for those that don't care for the gzip route. 

+1 on bumping this issue to a 1.1 consideration, just in the interest of finalizing this first draft spec.

as long as we there is one compelling case in which the index approach is better and one compelling case in which the full data is better (and I believe there are), it seems reasonable and not much effort to support both.

**Comment by edjez** [comment]
Resolved to postpone from 1.1/current release
